### PR TITLE
GPU/DMA: Flush the source region and invalidate the destination region when doing a DMA transfer.

### PR DESCRIPTION
**Pull request by Subv**
Also implemented the source offsets for Tiled->Linear transfers, this is needed by nouveau for proper glReadPixels behavior.

This should probably fix some random cache issues in some games. The same flushing for the 2D engine and the Copy engine will be implemented in a future PR.

This causes crashes in Splatoon, BoTW and ARMS due to our GetSwizzleOffset function being slightly broken:

```
GetSwizzleOffset(848, 1024, 1920, 4, 16) = 0x7EA000 for a 1920x1080 texture with bpp=4
That's out of bounds
```

### CODE REVIEW

**Comment by bunnei** [comment]
Causes some bugs in One Piece:

image

Also causes some big perf regressions:

image

File: src/video_core/renderer_opengl/gl_rasterizer.cpp [hunk 1]
```diff
@@ -537,9 +537,19 @@ void RasterizerOpenGL::DrawArrays() {
...
+    shader_cache.FlushRegion(0, Kernel::VMManager::MAX_ADDRESS);
+    buffer_cache.FlushRegion(0, Kernel::VMManager::MAX_ADDRESS);
+}
 
-void RasterizerOpenGL::FlushRegion(VAddr addr, u64 size) {}
+void RasterizerOpenGL::FlushRegion(VAddr addr, u64 size) {
+    MICROPROFILE_SCOPE(OpenGL_CacheManagement);
+    res_cache.FlushRegion(addr, size);
+    shader_cache.FlushRegion(addr, size);
+    buffer_cache.FlushRegion(addr, size);
```
Comment (degasus) [inline src/video_core/renderer_opengl/gl_rasterizer.cpp]: I wonder, will the compiler drop the calls to the shader_cache and the buffer_cache? If not, we should not call it here. This is a very often called function, and every overhead here matters.

File: src/video_core/engines/maxwell_dma.cpp [hunk 2]
```diff
@@ -44,37 +46,67 @@ void MaxwellDMA::HandleCopy() {
...
+
+        // Copy the data to a staging buffer first to make applying the src and dst offsets easier
+        std::vector<u8> staging_buffer(regs.src_pitch * regs.src_params.size_y);
+
+        // In this mode, the src_pitch register contains the source stride, and the dst_pitch
+        // contains the bytes per pixel.
+        u32 src_bytes_per_pixel = regs.src_pitch / regs.src_params.size_x;
+        u32 dst_bytes_per_pixel = regs.dst_pitch;
+
+        Texture::CopySwizzledData(regs.src_params.size_x, regs.src_params.size_y,
```
Comment (FernandoS27) [inline src/video_core/engines/maxwell_dma.cpp]: You might want to use FastSwizzle here when applicable.

File: src/video_core/engines/maxwell_dma.cpp [hunk 3]
```diff
@@ -44,37 +46,67 @@ void MaxwellDMA::HandleCopy() {
...
+        ASSERT(regs.src_params.pos_x == 0);
+        ASSERT(regs.src_params.pos_y == 0);
+
         // If the input is linear and the output is tiled, swizzle the input and copy it over.
         Texture::CopySwizzledData(regs.dst_params.size_x, regs.dst_params.size_y, 1, 1, dst_buffer,
                                   src_buffer, false, regs.dst_params.BlockHeight());
     }
+
+    // We have to invalidate the destination region to evict any outdated surfaces from the cache.
+    rasterizer.InvalidateRegion(dest_cpu, copy_size);
```
Comment (degasus) [inline src/video_core/engines/maxwell_dma.cpp]: Please move this to the `rasterizer.FlushRegion(source_cpu, copy_size);` call.

File: src/video_core/renderer_opengl/gl_rasterizer.cpp [hunk 4]
```diff
@@ -351,6 +363,12 @@ void RasterizerOpenGL::ConfigureFramebuffers(bool using_color_fb, bool using_dep
     }
 
     if (depth_surface) {
+        if (depth_surface) {
```
Comment (degasus) [inline src/video_core/renderer_opengl/gl_rasterizer.cpp]: What is the point of the second if clause here?

File: src/video_core/renderer_opengl/gl_rasterizer.cpp (hunk 1 above)
Comment (Subv) [reply src/video_core/renderer_opengl/gl_rasterizer.cpp]: I checked the code generated by MSVC, it does remove the calls to the other functions.

I imagine if MSVC's optimizer does it, clang's and gcc's should do it as well (I don't have a way to test those)

File: src/video_core/renderer_opengl/gl_rasterizer.cpp (hunk 4 above)
Comment (Subv) [reply src/video_core/renderer_opengl/gl_rasterizer.cpp]: Copy-paste error

**Comment by BSoDGaming** [comment]
This causes some issue but fixes many things. Fixes first

One Piece With this PR
overlay issue fixed

Without this PR
overlay issue not fixed

Octopath also has much better graphics/fonts rendered in menus (Especially so with accurate framebuffers set to ON)
octo accurate framebuffers on

However it also introduces graphical errors on 3D models in One Piece
one piece face issue

face lines

It also causes BOTW to close Yuzu after selecting a save and arms closes in Yuzu before the main menu also

**Comment by ReaperOfSouls1909** [comment]
https://drive.google.com/drive/folders/1nJOH8s7q9j-tyDC5nynl1hePnm9yVaF0 crash logs for the newest parts of this PR almost all the games are broken and crash

File: src/video_core/renderer_opengl/gl_rasterizer_cache.cpp [hunk 5]
```diff
@@ -275,9 +275,12 @@ void MortonCopy(u32 stride, u32 block_height, u32 height, u8* gl_buffer, std::si
         const std::size_t size_to_copy{std::min(gl_buffer_size, data.size())};
         memcpy(gl_buffer, data.data(), size_to_copy);
     } else {
+        std::vector<u8> data(height * stride * bytes_per_pixel);
```
Comment (FernandoS27) [inline src/video_core/renderer_opengl/gl_rasterizer_cache.cpp]: stride = width*bytes_per_pixel
h = ((height + 8 - 1)/8)*8;
s = ((stride + 64 - 1)/64)*64;
size_to_copy = h*s;

**Comment by bunnei** [comment]
Superseded by #1517, #1497, and #1446.

**Pull request closed without merging**